rust/lance/src/dataset.rs

```diff
@@ -1436,7 +1438,216 @@ impl Dataset {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct BatchInfo {
+    pub fragment_id: u32,
+    pub batch_index: usize,
+}
+
+pub trait UDFCache: Send + Sync {
+    fn get_batch(&self, info: &BatchInfo) -> Result<Option<RecordBatch>>;
+    fn insert_batch(&self, info: BatchInfo, batch: RecordBatch) -> Result<()>;
+    fn get_fragment(&self, fragment_id: u32) -> Result<Option<Fragment>>;
+    fn insert_fragment(&self, fragment: Fragment) -> Result<()>;
+}
+
+pub struct BatchUDF {
+    #[allow(clippy::type_complexity)]
+    pub mapper: Box<dyn Fn(&RecordBatch) -> Result<RecordBatch> + Send + Sync>,
+    /// The schema of the returned RecordBatch
+    pub append_schema: Arc<ArrowSchema>,
```

Comment:
Lets call it `output_schema` to be consistent with the python code above, also this UDF struct can be more generic.